ledger/sandbox/src/main/scala/com/digitalasset/platform/index/LedgerBackedIndexService.scala

```diff
@@ -135,70 +128,47 @@ abstract class LedgerBackedIndexService(
-      case LedgerOffset.Absolute(absBegin) =>
-        endAt
-          .map(converter.toAbsolute(_).map(Some(_)))
-          .getOrElse(Source.single(None))
-          .flatMapConcat { endOpt =>
-            lazy val stream =
-              ledger.ledgerEntries(Some(absBegin.toLong), endOpt.map(_.value.toLong))
-
-            val finalStream = endOpt match {
-              case None => stream
-
-              case Some(LedgerOffset.Absolute(`absBegin`)) =>
-                Source.empty
-
-              case Some(LedgerOffset.Absolute(end)) if absBegin.toLong > end.toLong =>
-                Source.failed(
-                  ErrorFactories.invalidArgument(s"End offset $end is before Begin offset $begin."))
-
-              case Some(LedgerOffset.Absolute(end)) =>
-                val endL = end.toLong
-                stream
-                  .takeWhile(
-                    {
-                      case (offset, _) =>
-                        //note that we can have gaps in the increasing offsets!
-                        (offset + 1) < endL //api offsets are +1 compared to backend offsets
-                    },
-                    inclusive = true // we need this to be inclusive otherwise the stream will be hanging until a new element from upstream arrives
-                  )
-                  // after the following step, we will add +1 to the offset before we expose it on the ledger api.
-                  // when the interval is [5, 7[, then we should only emit ledger entries with _backend_ offsets 5 and 6,
-                  // which then get changed to 6 and 7 respectively. the application can then take the offset of the last received
-                  // transaction and use it as a begin offset for another transaction service request.
-                  .filter(_._1 < endL)
-            }
-            // we MUST do the offset comparison BEFORE collecting only the accepted transactions,
-            // because currentLedgerEnd refers to the offset of the mixed set of LedgerEntries (e.g. completions, transactions, ...).
-            // If we don't do this, the response stream will linger until a transaction is committed AFTER the end offset.
-            // The immediate effect is that integration tests will not complete within the timeout.
-            finalStream.collect {
-              case (offset, t: LedgerEntry.Transaction) =>
-                (LedgerOffset.Absolute(LedgerString.assertFromString((offset + 1).toString)), t)
-            }
-          }
+    converter.toAbsolute(beginExclusive).flatMapConcat { begin =>
+      endInclusive
+        .map(converter.toAbsolute(_).map(Some(_)))
+        .getOrElse(Source.single(None))
+        .flatMapConcat {
+          case Some(`begin`) =>
+            Source.empty
+
+          case Some(end) if begin > end =>
+            Source.failed(
+              ErrorFactories.invalidArgument(s"End offset $end is before Begin offset $begin."))
+
+          case endOpt @ (None | Some(_)) =>
```

Comment:
Any reason to not simply make this a wildcard? The pattern describes any `Option` anyway. :slightly_smiling_face:

Reply:
I like to avoid wildcard matches whenever possible. Makes it slightly safer when refactoring, because you see all the branches explicitly. This applies in general, but might not be such an issue in this particular case for matching on `Option`.

Reply:
Makes sense, thanks.